Edit mode/Action Bar updates

charliescheer opened the pull request:
### Fix
Fixes #1326 

This PR removes the Action bar from Simplenote iOS.  This has affects the new note button and edit mode.  

Edit mode has been moved into the long press context menu
<img height="400px" src="https://cdn-std.droplr.net/files/acc_593859/rApGoX" />

when in edit mode the new note button is replaced with the Cancel button.  When edit mode is exited the new note button returns.

This also puts the new note button back in the top navigation bar
<img height="400px" src= "https://cdn-std.droplr.net/files/acc_593859/Dyow50" />


### Test
1. Load the app and confirm that in the note list and the editor you don't see the bottom bar and that the new note button is in the top right navigation toolbar
2. Long press on a note from the note list  to see the select option
3. tap on the select option.  Notice the new note button changes to cancel.  
4. Try canceling and select/deleting to make sure the new note button returns after edit mode is exited

### Review
***(Required)*** Add instructions for reviewers.  For example:
> Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release

> `RELEASE-NOTES.txt` was updated in 4e5ad8 with:
> 
> > -   Moved edit mode into a long press action and removed the new note bar #1328

charliescheer committed:
Add edit mode to context menu

charliescheer committed:
Moved new note button back to top in list mode

charliescheer committed:
Move new note button back to top in editor

charliescheer committed:
Added new icon image for select mode

charliescheer committed:
Removed some lingering iPad only exclusions for new note bar

charliescheer committed:
Fixed a bug where the select icon wasn't changing color in dark mode

charliescheer committed:
Updated release notes

eshurakov reviewed (approved):
Looks good to me!

Questions:
* should we preselect the long-tapped note after user chooses "select"?
* should we disable "Delete" button if nothing is selected?

cc @SylvesterWilmott

pachlava commented:
> should we preselect the long-tapped note after user chooses "select"?

This was my assumption here as well, I would expect this as a user

> should we disable "Delete" button if nothing is selected?

this is tracked in #1308

Generally, all issues that were created for the edit mode: #1306, #1307, #1308, #1311 are still applicable here

pachlava commented:
Tried this scenario (assuming that tapping `Select Notes` should auto-select the note):

1. Have so many notes that scrolling the list is needed
2. Scroll to the very bottom
3. Select the last note ~> this note is now hidden behind the bottom bar

https://user-images.githubusercontent.com/73365754/122200700-f4facb80-cea3-11eb-8a2d-eece4fcdd42e.mov

I would expect that just selected note should always be visible on screen.

SylvesterWilmott commented:
Looks good!

In reference to what @eshurakov mentioned about pre-selection, the label in the contextual menu should be "Select" instead of "Select notes" and therefore indicate that the current note will be selected. So, you should always enter edit mode with one note selected.

charliescheer committed:
Changed enabling select to select cell that was pressed on

charliescheer committed:
Disabled the trash button when there are no selected notes

charliescheer commented:
Thanks for the feedback y'all.  I have fixed the two issues raised:

>should we preselect the long-tapped note after user chooses "select"?
>should we disable "Delete" button if nothing is selected?

I also fixed a few of the things mentioned in #1308 

I looked into #1307 #1306 and #1311 I will need a bit more time to dig into those, I think it is better to exclude those from this PR, but it is on my radar to circle back to those soon

charliescheer commented:
@SylvesterWilmott when you get a chance can you take another look at the selecting behavior, it will now select the note that was long pressed when entering edit mode

charliescheer committed:
removed testing print statement

pachlava commented:
Did a bit of testing with recent additions - looks great! One question we can discuss here maybe: should edit mode be stopped once note is deleted via edit mode? This is what happens now.

cc @SylvesterWilmott

SylvesterWilmott commented:
Look great!

>should edit mode be stopped once note is deleted via edit mode?

That's the expected behaviour yes, based on what generally happens in system selections of this type.

One last thing, unless there is a strong objection to it. Can we change the label "Select Notes" to "Select"? I think this better implies that the current note will be selected after pressing.

charliescheer committed:
Changed search interaction to disabled when edit active

charliescheer committed:
Changed title for select notes in context menu to select

charliescheer commented:
Those changes have been made, search is now disabled when editing and the menu item is now "Select" @SylvesterWilmott can you take another look when you have a chance

@eshurakov can you also take another look at how I am disabling the search bar.  It is a fairly straight forward change, but I wasn't sure if we wanted to disable the search bar directly or use the search bar controller (I used the search bar controller to manage disabling the search)

TY!!

SylvesterWilmott commented:
Looks great, thanks @charliescheer .

A couple things:

Since the search bar has some degree of transparency already and therefore blurred elements show through it, can we instead only apply transparency to the search placeholder label while search is disabled? So to clarify. When the search bar is disabled, placeholder text goes to 0.5 opacity.

When searching, you are able to enter edit mode which then results in not being able to exit. I think at this time, removing or disabling the "Select" option from the contextual menu while in search is fine. There are plans to move the tag search out of the main search results view which will have an effect on how this works and will need to be defined in the design too.

eshurakov commented on Simplenote/Classes/SearchDisplayController.swift:
```diff
@@ -79,6 +79,11 @@ class SearchDisplayController: NSObject {
         updateNavigationBar(hidden: active)
     }
+
+    func setEnabled(_ enabled: Bool) {
+        searchBar.isUserInteractionEnabled = enabled
+        searchBar.alpha = enabled ? UIKitConstants.alpha1_0 : UIKitConstants.alpha0_5
+    }
```
I think this is good 👍 

We may also want to disable search activation via shortcut when in edit mode, right now it is possible.

charliescheer replied:
Good suggestion.  Disabled

eshurakov commented on Simplenote/Classes/SPNoteListViewController+Extensions.swift:
```diff
@@ -830,14 +840,20 @@ private extension SPNoteListViewController {
 
+        let select = UIAction(title: ActionTitle.select, image: .image(name: .success)) { [weak self] _ in
+            self?.setEditing(true, animated: true)
+            self?.tableView.selectRow(at: indexPath, animated: false, scrollPosition: .bottom)
```
I checked with @SylvesterWilmott and we don't want to always scroll selected row to the bottom.

I guess that you wanted to fix this issue: https://github.com/Automattic/simplenote-ios/pull/1328#issuecomment-862232153 and I would expect `scrollPosition: .none` to work, but for some reason it doesn't :-(

I tried a few things and the following combo worked, it only scrolls the row if it is covered by the bottom toolbar, otherwise scroll position is kept intact

```
self?.tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
self?.tableView.scrollToNearestSelectedRow(at: .none, animated: false)
```

charliescheer replied:
Great.  TY!

charliescheer committed:
Changed alpha for search bar placeholder text when vc is editing

charliescheer committed:
Disabled edit mode when search is active

charliescheer committed:
Disabled search keyboard shortcut when note list is editing

charliescheer committed:
Fixed a bug where entering edit mode might scroll to the wrong place

charliescheer commented:
@SylvesterWilmott @eshurakov Thanks for the suggestions!  I have updated the PR to reflect those changes.  When you have a chance this is ready to be reviewed again

eshurakov reviewed (approved):
Looks good to me 👍

AliSoftware committed:
Update fastlane & fastlane-plugin-appcenter

SylvesterWilmott commented:
Looks great @charliescheer thanks!

pachlava commented:
Some additions from my side:

1. When we select notes, the navbar title usually changes to 'X Selected', where X is the number of selected notes. But this is not working for the very first selection. In the video you can see that with the 1st selected note, the title is still 'All Notes':

https://user-images.githubusercontent.com/73365754/122923445-6687bd80-d36d-11eb-8d4e-a14430c65533.mov

2. Taking iPad+keyboard setup possibility into account, I think it's worth disabling the new note shortcut, to avoid weird situations. Generally, shortcuts are very unreliable when used for Simulator, because of this I could reproduce new note creation from the edit mode only once. As for search shortcut, I did not reproduce it, either because it's disabled, or because it was not properly handled by Simulator.

charliescheer committed:
Fixed a bug where first note in selection mode would not change title

charliescheer committed:
Disabled the new note keyboard shortcut when in edit mode

charliescheer committed:
Merge branch 'issue/1326-edit-mode-update' of https://github.com/Automattic/simplenote-ios into issue/1326-edit-mode-update

charliescheer commented:
@pachlava thanks for those suggestions!  I updated the PR to fix those issues

charliescheer committed:
Merge branch 'develop' into issue/1326-edit-mode-update

mokagio commented:
The failures in both UI tests CI jobs are due to know issues that we're trying to iron out. It's actually possible that if you do another merge of `develop` into this branch (or even better _IMHO_ rebase this onto `develop` to keep the history linear) the tests will pass thanks to @pachlava's changes from #1335.

I say let's just merge this in, though, so it's in before the 4.39 code freeze.
